mixer/adapter/cloudwatch/logHandler.go

```diff
@@ -0,0 +1,158 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY Type, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cloudwatch
+
+import (
+	"html/template"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/cloudwatchlogs"
+	multierror "github.com/hashicorp/go-multierror"
+
+	"istio.io/istio/mixer/pkg/pool"
+	"istio.io/istio/mixer/template/logentry"
+)
+
+const (
+	// Cloudwatchlogs enforced limit on number of logentries per request
+	// https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_PutLogEvents.html
+	batchcount = 10000
+
+	// A default template for log entries will be used if operator doesn't supply one
+	defaultTemplate = `{{or (.source_ip) "-"}} - {{or (.source_user) "-"}} ` +
+		`[{{or (.timestamp.Format "2006-01-02T15:04:05Z07:00") "-"}}] "{{or (.method) "-"}} {{or (.url) "-"}} ` +
+		`{{or (.protocol) "-"}}" {{or (.response_code) "-"}} {{or (.response_size) "-"}}`
+)
+
+func (h *handler) generateLogEntryData(insts []*logentry.Instance) []*cloudwatchlogs.InputLogEvent {
+	logentryData := make([]*cloudwatchlogs.InputLogEvent, 0, len(insts))
+
+	for _, inst := range insts {
+		// cloudwatchlogs accepts unix timestamp in milliseconds
+		timestamp := inst.Timestamp.UnixNano() / int64(time.Millisecond)
+
+		message, err := getMessageFromVariables(h, inst)
+		if err != nil {
+			h.env.Logger().Errorf("failed to get message for instance: %s. %v", inst.Name, err)
+			continue
+		}
+
+		logEvent := cloudwatchlogs.InputLogEvent{
+			Message:   aws.String(message),
+			Timestamp: aws.Int64(timestamp),
+		}
+
+		logentryData = append(logentryData, &logEvent)
+	}
+	return logentryData
+}
+
+func getMessageFromVariables(h *handler, inst *logentry.Instance) (string, error) {
+	tmplString := h.cfg.GetLogs()[inst.Name].PayloadTemplate
+	tmpl, err := template.New(inst.Name).Parse(tmplString)
```

Comment:
does it make sense to allocate the template and parse the string in the builder?

Reply:
I don’t fully understand the comment. I parsed the template based on the info here https://golang.org/pkg/text/template/ . Can you please elaborate how else could I have done.

Reply:
I was wondering if it would make sense to do the `template.New(inst.Name).Parse(tmplString)` bits in the builder up front.

Since the template must be in the config (or use the default), you could do the parsing up front once, instead of for each call to `getMessageFromVariables`.

Then in `getMessageFromVariables` you could just do something like: `tmpl := h.template(inst.Name)` where template is something like:

```
func (h *handler) template(name string) text.Template {
  if t, found := h.templates[name]; found {
     return t
  }
  return h.defaultTemplate
}
```

Reply:
I have made the changes in the PR. Please review.